Dockerize heartbeat

tcnichol opened the pull request:
I'm leaving this one as draft. I have moved the extractor heartbeat stuff to its own folder alongside backend and frontend, with the idea that it will run as a separate service.

This means that inside of that folder I am copying some of the code from backend. I have also added this in the docker-compose.dev.yml file, but so far it does not appear that it connects to rabbitmq from inside there. It does connect to rabbitmq when it is run outside of docker. 

Could use some feedback on this.

lmarini commented:
@tcnichol @max-zilla I think we can move `heartbeat_listener_sync.py` to the main `app` folder and have a separate Dockerfile for it that only includes that file and required dependencies. This will avoid duplicating code and make sure that the dependencies are always in sync.

max-zilla commented:
I did a little code cleanup and merged in main - @tcnichol please give it another try, it looks like it is creating exchange and listening to RMQ properly now, but I didn't test sending actual heartbeat.

Also agree with Luigi about not needing a separate folder with repeated models etc.

tcnichol commented:
I'll move it back inside the app folder and then I will test.

tcnichol commented:
I'm currently getting this error when I try to run the docker container:

`  File "/code/./app/heartbeat_listener_sync.py", line 8, in <module>
    from models.extractors import (
  File "/code/app/models/extractors.py", line 4, in <module>
    from app.models.mongomodel import MongoModel
ModuleNotFoundError: No module named 'app'`

tcnichol commented:
I merged main into this branch since the extractor registration is now fixed on main. 

I don't think I have permission to push the docker image though, but I also didn't know if I should push it until it's tested and we know it works. You can build a local docker image by doing

`docker build -f heartbeat.Dockerfile -t  clowder/clowder2-heartbeat:latest  .
`

lmarini reviewed (changes requested):
ExtractorHeartbeat and Backend Lint actions are failing.

lmarini reviewed (changes requested):
Getting a runtime error when running the container after building it with `docker compose up --build`.

```2023-01-12 11:19:08 Traceback (most recent call last):
2023-01-12 11:19:08   File "/code/app/heartbeat_listener_sync.py", line 7, in <module>
2023-01-12 11:19:08     from app.config import settings
2023-01-12 11:19:08 ModuleNotFoundError: No module named 'app.config'```

tcnichol commented on backend/app/heartbeat_listener_sync.py:
```diff
@@ -0,0 +1,74 @@
+import logging
```
Ill get these added this afternoon.

tcnichol commented on backend/Pipfile:
```diff
@@ -23,6 +23,9 @@ elasticsearch = "*"
 pipenv = "*"
 install = "*"
 rocrate = "*"
+packaging= "*"
```
I don't think these are used, and am not sure how they got in. Removing them now.

tcnichol replied:
I added the packaging module since it parses versions - I am adding it back in since I do use it.

tcnichol commented:
I fixed the runtime error. Added PYTHONPATH to the heartbeat.Dockerfile and now the imports are working.

There seem to be new build errors now, looks like when doing the pytest it can't connect to keycloak.
